README.md
<img src="https://raw.githubusercontent.com/ADMAntwerp/CounterPlots/main/_static/counterplots_logo.svg"><br>

--------------------------------------

CounterPlots: Plotting tool for counterfactuals
=======================================

[![PyPI Version](https://img.shields.io/pypi/v/counterplots.svg)](https://pypi.org/project/counterplots/)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![example workflow](https://github.com/ADMAntwerp/CounterPlots/actions/workflows/deployment.yml/badge.svg)](https://github.com/ADMAntwerp/CounterPlots/actions)
[![Code Coverage](https://codecov.io/gh/rmazzine/counterplotcoverage/branch/main/graph/badge.svg?token=TQYJSGEMP1)](https://codecov.io/gh/rmazzine/counterplotcoverage)
[![Known Vulnerabilities](https://snyk.io/test/github/ADMAntwerp/CounterPlots/badge.svg)](https://snyk.io/test/github/ADMAntwerp/CounterPlots)

Counterplots is a Python package that allows you to plot counterfactuals with easy integration with any counterfactual generation algorithm.

[Medium Post with full tutorial](https://medium.com/@mazzine/measuring-counterfactual-explanations-feature-importance-with-counterplots-41359bc85aea)

[Preprint Article](https://arxiv.org/abs/2306.06506)

## Plot examples

### Greedy Plot

The greedy plot shows the greediest (feature change with the highest impact towards the opposite class) path from the factual instance until it reaches the counterfactual.

<img src="https://raw.githubusercontent.com/ADMAntwerp/CounterPlots/main/_static/counterplots_example_greedy.png">

### CounterShapley Plot

This chart shows each counterfactual feature change contribution to the counterfactual prediction. It uses Shapley values to calculate the contribution of each feature change.

<img src="https://raw.githubusercontent.com/ADMAntwerp/CounterPlots/main/_static/counterplots_example_countershapley.png">

### Constellation Plot

This chart shows the prediction score change for all possible feature change combinations.

<img src="https://raw.githubusercontent.com/ADMAntwerp/CounterPlots/main/_static/counterplots_example_constellation.png">

## Requirements
CounterPlots requires Python 3.8 or higher.

## Installation
With pip:
```bash
pip install counterplots
```

## Usage
To use CounterPlots, you just need the machine learning model predictor, and the factual and counterfactual points.
The example below uses a simple mock model:
```python
from counterplots import CreatePlot
import numpy as np

# Simple mock model for the predict_proba function which returns a probability for each input instance
def mock_predict_proba(data):
    out = []
    for x in data:
        if list(x) == [0.0, 0.0, 0.0]:
            out.append(0.0)
        elif list(x) == [1.0, 0.0, 0.0]:
            out.append(0.44)
        elif list(x) == [0.0, 1.0, 0.0]:
            out.append(0.4)
        elif list(x) == [0.0, 0.0, 1.0]:
            out.append(0.2)
        elif list(x) == [1.0, 1.0, 0.0]:
            out.append(0.3)
        elif list(x) == [0.0, 1.0, 1.0]:
            out.append(0.25)
        elif list(x) == [1.0, 0.0, 1.0]:
            out.append(0.4)
        elif list(x) == [1.0, 1.0, 1.0]:
            out.append(1.0)
    return np.array(out)

# Factual Instance
factual = np.array([0, 0, 0])
# Counterfactual Instance
cf = np.array([1, 1, 1])

# Create the plot object
cf_plots = CreatePlot(
    factual,
    cf,
    mock_predict_proba)

# Create the greedy plot
cf_plots.greedy('greedy_plot.png')
# Create the countershapley plot
cf_plots.countershapley('countershapley_plot.png')
# Create the constellation plot
cf_plots.constellation('constellation_plot.png')

# Print the countershapley values
print(cf_plots.countershapley_values())
```

In case you want to add custom names to the features, use the optional argument `feature_names`:
```python
cf_plots = CreatePlot(
    factual,
    cf,
    mock_predict_proba,
    feature_names=['feature1', 'feature2', 'feature3'])
```

In case you want to add custom labels to the factual and counterfactual points, use the optional argument `class_names`:
```python
cf_plots = CreatePlot(
    factual,
    cf,
    mock_predict_proba,
    class_names=['Factual', 'Counterfactual'])
```

## Using with Scikit-Learn

CounterPlots can be used with any machine learning model that has a `predict_proba` function. For example, with Scikit-Learn:
<details>

```python
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import load_iris

from counterplots import CreatePlot

iris = load_iris()

X = iris.data
y = [0 if l == 0 else 1 for l in iris.target] # Makes it a binary classification problem

clf = RandomForestClassifier(max_depth=2, random_state=0)
clf.fit(X, y)

preds = clf.predict(X)

# For the factual point, takes an instance with 0 classification
factual = X[np.argwhere(preds == 0)[0]][0]
# For the counterfactual point, takes an instance with 1 classification
cf = X[np.argwhere(preds == 1)[0]][0]

cf_plots = CreatePlot(
    factual,
    cf,
    clf.predict_proba,
    feature_names=iris.feature_names,
    class_names={0: 'Setosa', 1: 'Non-Setosa'}
)


# Create the greedy plot
cf_plots.greedy('iris_greedy_plot.png')
# Create the countershapley plot
cf_plots.countershapley('iris_countershapley_plot.png')
# Create the constellation plot
cf_plots.constellation('iris_constellation_plot.png')

# Print the countershapley values
print(cf_plots.countershapley_values())
```
</details>

## Citation
If you use CounterPlots in your research, please cite the following paper:
```bibtex
@misc{meulemeester2023calculating,
      title={Calculating and Visualizing Counterfactual Feature Importance Values}, 
      author={Bjorge Meulemeester and Raphael Mazzine Barbosa De Oliveira and David Martens},
      year={2023},
      eprint={2306.06506},
      archivePrefix={arXiv},
      primaryClass={cs.LG}
}
```


HISTORY.md
# 0.0.7 / 2023-06-10
==================
* Upgrade dependencies requirements
# 0.0.6 / 2023-06-10
==================
* Fix dependency issue
# 0.0.5 / 2023-06-10
==================
* Use the most recent packages possible
# 0.0.4 / 2023-06-10
==================
* Update dependency package versions
# 0.0.3 / 2023-06-10
==================
* Fix installation issue
# 0.0.2 / 2023-06-10
==================
* Updated documentation
# 0.0.1 / 2023-06-10
==================
* Initial package release

tests/test____init__.py
import numpy as np
import unittest
from unittest.mock import patch
from counterplots import CreatePlot


class CreatePlotTests(unittest.TestCase):

    def setUp(self):
        # Create dummy data for testing
        self.factual = np.array([1, 2, 3])
        self.cf = np.array([4, 5, 6])
        self.model_pred = lambda x: np.array([np.array([0]) if sum(r) < 10 else np.array([1]) for r in x])
        self.feature_names = ['feat1', 'feat2', 'feat3']
        self.class_names = {0: '0', 1: '1'}

    def test_init(self):
        create_plot = CreatePlot(self.factual, self.cf, self.model_pred, self.feature_names, self.class_names)
        self.assertTrue(np.array_equal(create_plot.factual, self.factual))
        self.assertTrue(np.array_equal(create_plot.cf, self.cf))
        self.assertEqual(create_plot.feature_names, self.feature_names)
        self.assertEqual(create_plot.class_names, self.class_names)
        self.assertEqual(create_plot.factual_score, 0)
        self.assertEqual(create_plot.cf_score, 1)

    def test_init_no_feat_name(self):
        create_plot = CreatePlot(self.factual, self.cf, self.model_pred, None, self.class_names)
        self.assertTrue(np.array_equal(create_plot.factual, self.factual))
        self.assertTrue(np.array_equal(create_plot.cf, self.cf))
        self.assertEqual(create_plot.feature_names, ['f1', 'f2', 'f3'])
        self.assertEqual(create_plot.class_names, self.class_names)
        self.assertEqual(create_plot.factual_score, 0)
        self.assertEqual(create_plot.cf_score, 1)

    def test_init_no_class_name(self):
        create_plot = CreatePlot(self.factual, self.cf, self.model_pred, self.feature_names, None)
        self.assertTrue(np.array_equal(create_plot.factual, self.factual))
        self.assertTrue(np.array_equal(create_plot.cf, self.cf))
        self.assertEqual(create_plot.feature_names, self.feature_names)
        self.assertEqual(create_plot.class_names, {0: '0', 1: '1'})
        self.assertEqual(create_plot.factual_score, 0)
        self.assertEqual(create_plot.cf_score, 1)

    @patch('counterplots.make_greedy_plot')
    @patch('counterplots.greedy_strategy')
    def test_greedy(self, mock_greedy_strategy, mock_make_greedy_plot):
        create_plot = CreatePlot(self.factual, self.cf, self.model_pred, self.feature_names, self.class_names)
        create_plot.greedy(save_path=None)  # Just check if it runs without error

    @patch('counterplots.make_countershapley_plot')
    @patch('counterplots._convert_legendvalue')
    @patch('counterplots.calc_shapley_values_between')
    def test_countershapley(self, mock_calc_shapley_values_between, mock__convert_legendvalue,
                            mock_make_countershapley_plot):

        mock_calc_shapley_values_between.return_value = ((0.4, 0.2, 0.1), (0, 1, 2))

        create_plot = CreatePlot(self.factual, self.cf, self.model_pred, self.feature_names, self.class_names)
        create_plot.countershapley(save_path=None)  # Just check if it runs without error

    @patch('counterplots.calc_shapley_values_between')
    def test_countershapley_values(self, mock_calc_shapley_values_between):
        mock_calc_shapley_values_between.return_value = ((0.4, 0.2, 0.1), (0, 1, 2))
        create_plot = CreatePlot(self.factual, self.cf, self.model_pred, self.feature_names, self.class_names)
        values = create_plot.countershapley_values()
        expected_values = {
            'feature_names': self.feature_names,
            'feature_values': (0.4, 0.2, 0.1),
            'feature_indices': (0, 1, 2)
        }
        self.assertEqual(values, expected_values)

    @patch('counterplots.make_constellation_plot')
    def test_constellation(self, mock_make_constellation_plot):
        create_plot = CreatePlot(self.factual, self.cf, self.model_pred, self.feature_names, self.class_names)
        create_plot.constellation(save_path=None)  # Just check if it runs without error

    @patch('counterplots.make_constellation_plot')
    @patch('counterplots.CreatePlot.greedy')
    def test_constellation_few_changes(self, mock_greedy, mock_make_constellation_plot):
        create_plot = CreatePlot(np.array([1, 2, 3]), np.array([1, 2, 4]),
                                 self.model_pred, self.feature_names, self.class_names)
        create_plot.constellation(save_path=None)

        # Check if greedy is called
        mock_greedy.assert_called_once_with(None)

        # Check if constellation is not called
        mock_make_constellation_plot.assert_not_called()


tests/utils/test_plots.py
import unittest
from unittest.mock import patch

import numpy as np

from counterplots.utils.plots import make_greedy_plot, make_countershapley_plot, make_constellation_plot


class TestMakeGreedyPlot(unittest.TestCase):
    @patch('counterplots.utils.plots.plt')
    @patch('counterplots.utils.plots.mpl')
    def test_make_greedy_plot(self, mock_mpl, mock_plt):
        factual_score = 0.7
        features_data = [
            {'score': 0.4, 'name': 'Feature 1', 'factual': 'A', 'counterfactual': 'B'},
            {'score': 0.6, 'name': 'Feature 2', 'factual': 'X', 'counterfactual': 'Y'}
        ]
        class_names = {0: 'Class A', 1: 'Class B'}
        save_path = 'plot.png'

        make_greedy_plot(factual_score, features_data, class_names, save_path)


class TestMakeCounterShapleyPlot(unittest.TestCase):

    @patch('counterplots.utils.plots.plt')
    @patch('counterplots.utils.plots.mpl')
    def test_make_countershapley_plot(self, mock_mpl, mock_plt):
        factual_score = 0.0
        features_data = [{'x': 62.0,
                          'score': 0.4,
                          'name': 'Feature 1',
                          'factual': -4.0,
                          'counterfactual': -1.0},
                         {'x': 38.0,
                          'score': 0.6,
                          'name': 'Feature 2',
                          'factual': 1.0,
                          'counterfactual': -1.0}]
        class_names = {0: 'Class A', 1: 'Class B'}
        save_path = 'plot.png'

        make_countershapley_plot(factual_score, features_data, class_names, save_path)


class TestMakeConstellationPlot(unittest.TestCase):

    @patch('counterplots.utils.plots.plt')
    @patch('counterplots.utils.plots.mpl')
    def test_make_constellation_plot(self, mock_mpl, mock_plt):
        factual_score = 0.0
        single_points_chart = np.array([[0., 0.00519673],
                                        [1., 0.06478132],
                                        [2., 0.004789]])
        text_features = ['Feature 6 (-3.19627➜-0.3652)',
                         'Feature 8 (2.0093➜-3.15686)',
                         'Feature 9 (3.1568➜-2.41393)']
        mulitple_points_chart = np.array(
            [[np.array([0, 1]), 0.263131977249517],
             [np.array([0, 2]), 0.024619989978766642],
             [np.array([1, 2]), 0.24759475995893304]], dtype='object')
        mulitple_points_chart_y = [0.5, 1.0, 1.5]
        single_points = np.array([
            [6.00000000e+00, 5.19673099e-03],
            [8.00000000e+00, 6.47813170e-02],
            [9.00000000e+00, 4.78899690e-03]])
        class_names = {0: 'Class 1', 1: 'Class 0'}
        cf_score = 0.63
        point_to_pred = {0: 0.005196730993880675, 1: 0.06478131701518305, 2: 0.004788996904253137}
        save_path = 'plot.png'

        make_constellation_plot(factual_score, single_points_chart, text_features, mulitple_points_chart,
                                mulitple_points_chart_y, single_points, class_names, cf_score, point_to_pred,
                                save_path)


tests/utils/test_process.py
import unittest
import numpy as np
from counterplots.utils.process import create_adapted_model_and_class, greedy_strategy, factorial, \
    calc_adapted_instance, generate_contribution_graph, calc_shapley_values_between, _convert_legendvalue


class TestScript(unittest.TestCase):

    def setUp(self): # noqa
        def model(data):
            out = []
            for x in data:
                if x[0] == 0.1 and x[1] == 0.2 and x[2] == 0.3:
                    out.append(0.0)
                elif x[0] == 0.4 and x[1] == 0.5 and x[2] == 0.6:
                    out.append(1.0)
                elif x[0] == 0.1 and x[1] == 0.5 and x[2] == 0.6:
                    out.append(0.4)
                elif x[0] == 0.4 and x[1] == 0.2 and x[2] == 0.6:
                    out.append(0.2)
                elif x[0] == 0.4 and x[1] == 0.5 and x[2] == 0.3:
                    out.append(0.3)
                elif x[0] == 0.1 and x[1] == 0.2 and x[2] == 0.6:
                    out.append(0.25)
                elif x[0] == 0.4 and x[1] == 0.2 and x[2] == 0.3:
                    out.append(0.4)
                elif x[0] == 0.1 and x[1] == 0.5 and x[2] == 0.3:
                    out.append(0.44)
            return np.array(out)
        self.model = model
        self.factual = np.array([0.1, 0.2, 0.3])
        self.cf = np.array([0.4, 0.5, 0.6])

    def test_create_adapted_model_and_class(self):
        class_names = ['A', 'B']
        model_pred = self.model
        factual = self.factual

        adapted_model, adapted_class = create_adapted_model_and_class(factual, model_pred, class_names)

        self.assertTrue(callable(adapted_model))
        self.assertEqual(adapted_class[0], 'A')
        self.assertEqual(adapted_class[1], 'B')

    def test_create_adapted_model_and_class_B_A(self):
        class_names = ['A', 'B']
        model_pred = self.model
        factual = self.cf

        adapted_model, adapted_class = create_adapted_model_and_class(factual, model_pred, class_names)

        self.assertTrue(callable(adapted_model))
        self.assertEqual(adapted_class[0], 'B')
        self.assertEqual(adapted_class[1], 'A')

    def test_create_adapted_model_and_class_double_pred(self):
        class_names = ['A', 'B']
        model_pred = lambda data: [[1 - self.model([x])[0], self.model([x])[0]] for x in data] # noqa
        factual = self.factual

        adapted_model, adapted_class = create_adapted_model_and_class(factual, model_pred, class_names)

        self.assertTrue(callable(adapted_model))
        self.assertEqual(adapted_class[0], 'A')
        self.assertEqual(adapted_class[1], 'B')

    def test_create_adapted_model_and_class_double_pred_B_A(self):
        class_names = ['A', 'B']
        model_pred = lambda data: [[1 - self.model([x])[0], self.model([x])[0]] for x in data] # noqa
        factual = self.cf

        adapted_model, adapted_class = create_adapted_model_and_class(factual, model_pred, class_names)

        self.assertTrue(callable(adapted_model))
        self.assertEqual(adapted_class[0], 'B')
        self.assertEqual(adapted_class[1], 'A')

    def test_greedy_strategy(self):
        factual = self.factual
        cf = self.cf
        adapted_model = self.model
        feature_names = ['X', 'Y', 'Z']

        features_data = greedy_strategy(factual, cf, adapted_model, feature_names)

        self.assertEqual(features_data,
                         [{'score': 0.44, 'name': 'Y', 'factual': 0.2, 'counterfactual': 0.5},
                          {'score': 0.4, 'name': 'Z', 'factual': 0.3, 'counterfactual': 0.6},
                          {'score': 1.0, 'name': 'X', 'factual': 0.1, 'counterfactual': 0.4}])

    def test_factorial(self):
        self.assertEqual(factorial(0), 1)
        self.assertEqual(factorial(1), 1)
        self.assertEqual(factorial(5), 120)

    def test_calc_adapted_instance(self):
        instance = self.factual
        instance2 = self.cf
        indices = [0, 2]

        adapted_instance = calc_adapted_instance(instance, instance2, indices)

        self.assertEqual(adapted_instance, [0.4, 0.2, 0.6])

    def test_generate_contribution_graph(self):
        instance1 = self.factual
        instance2 = self.cf
        model = lambda x: np.array([0.7]) # noqa
        indices_feature_diff = [0, 2]
        class_of_interest_index = 0

        contribution_graph = generate_contribution_graph(instance1, instance2, model, indices_feature_diff,
                                                         class_of_interest_index)

        self.assertEqual(len(contribution_graph), 4)
        self.assertIn((), contribution_graph)
        self.assertIn((0,), contribution_graph)
        self.assertIn((2,), contribution_graph)
        self.assertIn((0, 2), contribution_graph)

    def test_calc_shapley_values_between(self):
        instance1 = self.factual
        instance2 = self.cf
        model = self.model

        class_of_interest_index = 0

        feature_contributions, indices = calc_shapley_values_between(
            instance1, instance2, model, class_of_interest_index)

        self.assertEqual(feature_contributions, (0.42166666666666663, 0.30166666666666664, 0.2766666666666666))
        self.assertEqual(indices, (1, 0, 2))

    def test__convert_legendvalue(self):
        self.assertEqual(_convert_legendvalue(10000), '1.00E4')
        self.assertEqual(_convert_legendvalue(0.00001), '1.00E-5')
        self.assertEqual(_convert_legendvalue(0.123456), 0.12346)
        self.assertEqual(_convert_legendvalue('ABC'), 'ABC')


exp_notebooks/exp_utils.py
import time
from nice import NICE
from counterplots import CreatePlot


class CounterPlotNice(NICE):

    def __init__(self, predict_fn, *args, **kwargs):
        super().__init__(predict_fn, *args, **kwargs)
        self.predict_fn = predict_fn

    def explain(self, X, feature_names=None, class_names=None, target_class='other'):
        explanations = super().explain(X, target_class)

        out_exp = []

        for i in range(len(X)):
            out_exp.append(CreatePlot(factual=X[i],
                                      cf=explanations[i],
                                      model_pred=self.predict_fn,
                                      feature_names=feature_names,
                                      class_names=class_names))

        return out_exp


class counterplot_explainer:

    def __init__(self, cf_generator, feature_names, class_names, model):
        self.cf_generator = cf_generator
        self.feature_names = feature_names
        self.class_names = class_names
        self.model = model

    def explain(self, factual):
        cf = self.cf_generator(factual)
        exp = CreatePlot(
            factual=factual,
            cf=cf,
            model_pred=self.model,
            feature_names=self.feature_names,
            class_names=self.class_names)

        results = exp
        results_cv = results.countershapley_values()
        r_fn = results_cv['feature_names']
        r_fv = results_cv['feature_values']
        results_per_idx = {self.feature_names.index(
            fn): fv for fn, fv in zip(r_fn, r_fv)}
        return results, results_per_idx


class shap_explainer:

    def __init__(self, explainer):
        self.explainer = explainer

    def explain(self, factual):
        exp = self.explainer(factual)
        results = exp
        results_per_idx = {idx: val for idx, val in enumerate(results[1])}

        return results, results_per_idx


class lime_explainer:

    def __init__(self, explainer, model):
        self.explainer = explainer
        self.model = model

    def explain(self, factual):
        exp = self.explainer.explain_instance(
            factual, self.model, num_features=len(factual))
        results = exp
        results_lv_idx = {feat[0]: feat[1] for feat in results.as_map()[1]}
        results_per_idx = {idx: results_lv_idx[idx]
                           for idx in range(len(factual))}
        return results, results_per_idx


def generate_explanations(data, exp_gen):
    gen_times = []
    results = []
    results_per_idx = []
    for i in range(len(data)):
        start_time = time.time()
        result, result_per_idx = exp_gen(data[i])
        gen_times.append(time.time() - start_time)
        results.append(result)
        results_per_idx.append(result_per_idx)

    return results, results_per_idx, gen_times


counterplots/utils/verification.py
import numpy as np


# Verifications
def verify_shape(array):
    if len(np.array(array).shape) != 1:
        raise ValueError('Array must be 1-dimensional.')


def verify_factual_counterfactual_shape(factual, cf):
    for type, data in zip(['Factual', 'Counterfactual'], [factual, cf]):
        if len(data.shape) != 1:
            raise ValueError(f'{type} array must be 1-dimensional. Got {data.shape}.')


def verify_factual_cf(factual, cf):
    if np.array(factual).shape != np.array(cf).shape:
        raise ValueError('Factual and counterfactual must have the same shape.')


def verify_feature_names(factual, feature_names):
    if len(feature_names) != np.array(factual).shape[0]:
        raise ValueError('Feature names must have the same length as the number of features.')


def verify_model_prediction(model_pred, factual):
    try:
        model_pred(np.array([factual]))
    except Exception as e:
        raise ValueError('Model prediction function must be callable but got error: {}'.format(e))


def verify_binary_class(model_pred, factual):
    model_pred_shape = model_pred(np.array([factual])).shape
    if len(model_pred_shape) == 1:
        return
    elif len(model_pred_shape) == 2:
        if model_pred_shape[1] > 2:
            raise ValueError('Model must be binary.')
    else:
        raise ValueError('Model must be binary.')


def verify_class_names(class_names):
    if type(class_names) != dict or len(class_names) != 2:
        raise ValueError(
            'Class names must be a dictionary, where int(0) is the class for prediction 0 and '
            'int(1) is the class for prediction 1.')


counterplots/utils/plots.py
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
from matplotlib.path import Path
import matplotlib.patches as patches


def make_greedy_plot(factual_score, features_data, class_names, save_path):
    fig = plt.figure(figsize=(7, 1 * len(features_data))) # noqa

    scatter_points = [f['score'] for f in features_data]
    scatter_points.insert(0, factual_score)

    plt.plot(scatter_points, range(len(scatter_points)),
             color='#c4c4c4', linestyle='dashed', zorder=0)
    for c_idx, point in enumerate(scatter_points):
        plt.scatter(
            [point],
            [c_idx],
            color='#ff0055D2' if point < 0.5 else '#008ae7D2',
            s=100)

    features_names = ['Factual']
    for feat_idx, f in enumerate(features_data):
        features_names.append(
            f'{f["name"]} ({f["factual"]}➜{f["counterfactual"]})')
    max_feat_names_length = max([len(feat_name)
                                for feat_name in features_names])

    # Plot a vertical line at the threshold score
    plt.axvline(x=0.5, color='#c20000', zorder=0, linewidth=0.5)

    # Plot a vertical line at the factual score
    plt.axvline(x=factual_score, color='#ff0055D2',
                linestyle='dashed', zorder=0)
    plt.text(factual_score - 0.088, len(features_names)*1.04 -
             0.9, 'Factual Score', color='#ff0055D2')

    # Plot a vertical line at the counterfactual score
    plt.axvline(x=scatter_points[-1],
                color='#008ae7D2', linestyle='dashed', zorder=0)
    plt.text(scatter_points[-1] - 0.15, len(features_names)*1.04 -
             0.9, 'Counterfactual Score', color='#008ae7D2')

    for feat_idx, feat_name in enumerate(features_names):
        plt.text(-0.02 * max_feat_names_length, feat_idx,
                 feat_name, color='#545454', fontsize=12)

    # Print binary class
    class_0_name = list(class_names.values())[0]
    class_1_name = list(class_names.values())[1]
    size_factual_class = mpl.textpath.TextPath(
        (0, 0), class_0_name).get_extents().width * 0.003
    plt.text(0.47 - size_factual_class - max([len(f) for f in features_names])*0.0001, len(features_names)
             - 0.75 + (len(features_names) + 1)*0.04,
             class_0_name, color='#ff0055D2', fontweight='bold')
    plt.text(0.49, len(features_names) - 0.75 + (len(features_names) + 1)*0.04, '➜',
             color='#c20000', fontweight='bold')
    plt.text(0.52, len(features_names) - 0.75 + (len(features_names) + 1)
             * 0.04, class_1_name, color='#008ae7D2', fontweight='bold')

    plt.gca().axes.get_yaxis().set_visible(False)
    plt.gca().spines['top'].set_visible(False)
    plt.gca().spines['left'].set_visible(False)
    plt.gca().spines['right'].set_visible(False)

    # Set limit of x axis from 0 to 1
    plt.xlim(-0.05, 1)

    if save_path:
        plt.savefig(save_path, bbox_inches='tight')


def make_countershapley_plot(factual_score, features_data, classes, save_path):
    fig = plt.figure(figsize=(10, 1.5))
    ax = fig.add_subplot(111)
    scale_x = 100
    scale_y = 100
    bar_y_height = scale_y / 2
    bar_y_height_sub = scale_y / 8.0
    fontsize = 10
    color_bar = '#48C975D2'
    cf_score = features_data[-1]['score']
    sum_countershapley = cf_score - factual_score

    # Beak must be always the 4%
    def create_bar(start_x, end_x, no_beak=False):

        verts = [
            (start_x, 0.),  # left, bottom
            (start_x, bar_y_height),  # left, top
            (end_x - 4, bar_y_height),  # right, top
            (end_x, bar_y_height),  # right, top
            (end_x, 0.),  # right, bottom
            (0., 0.),  # ignored
        ]

        codes = [
            Path.MOVETO,
            Path.LINETO,
            Path.LINETO,
            Path.LINETO if no_beak else Path.CURVE3,
            Path.LINETO,
            Path.CLOSEPOLY,
        ]

        return Path(verts, codes)

    prev_score = factual_score
    current_x = 0
    x_left_pos = []
    for _, feat_data in enumerate(features_data):
        x_left_pos.append(current_x)
        x_size = feat_data['x'] * 100 / scale_x + current_x

        # Plot text for feature name
        plt.text(
            current_x + (x_size - current_x) / 2 -
            len(feat_data['name']) * fontsize / 10 / 2,
            -40,
            feat_data['name'],
            color='#545454',
            fontsize=fontsize)

        feature_change_text = f"{feat_data['factual']}➜{feat_data['counterfactual']}"

        # Plot bar up to 50% of the plot with the factual color
        ax.add_patch(patches.PathPatch(create_bar(
            current_x, x_size), facecolor=color_bar, lw=0))

        # Plot text for feature changes
        feat_change_text = mpl.textpath.TextPath(
            (0, 0), feature_change_text, size=fontsize)
        plt.text(
            current_x + (x_size - current_x)/2 -
            feat_change_text.get_extents().width * 0.1,
            -80,
            feature_change_text,
            color='#545454',
            fontsize=fontsize)

        feat_cs_score_percentage = round(
            (feat_data['score'] - prev_score)/sum_countershapley*100, 1)
        feat_cs_score_percentage_text = mpl.textpath.TextPath(
            (0, 0), f'{feat_cs_score_percentage}%', size=fontsize)
        plt.text(
            current_x + (x_size - current_x) / 2 -
            feat_cs_score_percentage_text.get_extents().width * 0.085,
            bar_y_height_sub,
            f'{feat_cs_score_percentage}%',
            color='white',
            fontsize=fontsize,
            weight="bold")

        current_x = x_size
        prev_score = feat_data['score']

    # Print CF score
    plt.text(
        current_x,
        scale_y,
        round(features_data[-1]['score'], 2),
        color='#008ae7',
        fontsize=fontsize)

    # Draw bar for the factual score
    plt.bar(0, scale_y - 10, width=0.5, color='#ff0055', linewidth=1)
    plt.text(-5, scale_y + fontsize * 4, 'Factual Score',
             color='#ff0055', fontsize=fontsize)
    plt.text(0, scale_y, factual_score, color='#ff0055',
             fontsize=fontsize)

    # Plot bar for the counterfactual score
    plt.bar(current_x, scale_y - 10, width=0.5, color='#008ae7', linewidth=1)
    plt.text(current_x - 10, scale_y + fontsize * 4,
             'Counterfactual Score', color='#008ae7', fontsize=fontsize)

    # Remove y axis
    plt.gca().axes.get_yaxis().set_visible(False)
    plt.gca().axes.get_xaxis().set_visible(False)
    # Remove square around the plot
    plt.gca().spines['top'].set_visible(False)
    plt.gca().spines['left'].set_visible(False)
    plt.gca().spines['right'].set_visible(False)
    plt.gca().spines['bottom'].set_visible(False)

    for l_pos in x_left_pos:  # little dividers between x labels
        plt.bar(l_pos, -scale_y + 10, width=0.2, color='#cccccc', linewidth=0)
    plt.bar(scale_x, -scale_y + 10, width=0.2,
            color='#cccccc', linewidth=0)  # divider at the end

    ax.set_xlim(0, scale_x)
    ax.set_ylim(-scale_y, scale_y + 70)

    if save_path is not None:
        plt.savefig(save_path, bbox_inches='tight')

    plt.show()


def make_constellation_plot(factual_score, single_points_chart, text_features, mulitple_points_chart,
                            mulitple_points_chart_y, single_points, class_names, cf_score, point_to_pred, save_path):
    x_dim = 10
    y_dim = 4
    y_lim_low = -0.1
    y_lim_high = len(single_points_chart) - 0.9
    x_lim_low = 0
    x_lim_high = 1

    fig = plt.figure(figsize=(x_dim, y_dim))
    ax = fig.add_subplot(111)

    ax.set_xlim(x_lim_low, x_lim_high)
    ax.set_ylim(y_lim_low, y_lim_high)

    # Plot single change points
    for idx_p, p in enumerate(single_points_chart):
        plt.scatter([p[1]], [p[0]], color='#008ae7D2' if p[1]
                    >= 0.5 else '#ff0055D2', s=100)

    # Plot feature names and value changes
    max_text_features = max([len(f) for f in text_features])
    for i in range(len(text_features)):
        plt.text(-0.012*max_text_features, i,
                 text_features[i], color='#545454', fontsize=12)

    # Verify if there are multiple change points
    if len(mulitple_points_chart) > 0:
        # Plot multiple change points
        plt.scatter(mulitple_points_chart[:, 1],
                    mulitple_points_chart_y,
                    color=['#E0423A' if s <
                           0.5 else 'blue' for s in mulitple_points_chart[:, 1]],
                    s=10)

    # Plot counterfactual point
    cf_pred_x_1 = np.mean([*range(len(single_points))])
    plt.scatter([cf_score], [cf_pred_x_1], color='#008ae7D2' if cf_score
                >= 0.5 else '#ff0055D2', s=100)

    # Plot a vertical line at the threshold
    plt.axvline(x=0.5, color='#c20000', zorder=0, linewidth=0.5)

    # Plot a vertical line at the factual score
    plt.axvline(x=factual_score, color='#ff0055D2',
                linestyle='dashed', zorder=0)
    plt.text(factual_score - 0.06, len(text_features) *
             1.01 - 0.9, 'Factual Score', color='#ff0055D2')

    # Plot a vertical line at the counterfactual score
    plt.axvline(x=cf_score, color='#008ae7D2',
                linestyle='dashed', zorder=0)
    plt.text(cf_score - 0.10, len(text_features)*1.01 - 0.9,
             'Counterfactual Score', color='#008ae7D2')

    plt.gca().axes.get_yaxis().set_visible(False)
    plt.gca().spines['top'].set_visible(False)
    plt.gca().spines['left'].set_visible(False)
    plt.gca().spines['right'].set_visible(False)

    # Plot classes names
    class_0_name = class_names[0]
    class_1_name = class_names[1]
    size_factual_class = mpl.textpath.TextPath(
        (0, 0), class_0_name).get_extents().width * 0.002
    plt.text(0.48 - size_factual_class - max([len(f) for f in text_features])*0.0001,
             len(text_features) - 0.9 + len(text_features)*0.04, class_0_name, color='#ff0055D2', fontweight='bold')
    plt.text(0.49, len(text_features) - 0.9 + len(text_features)*0.04, '➜',
             color='#c20000', fontweight='bold')
    plt.text(0.51, len(text_features) - 0.9 + len(text_features)*0.04,
             class_1_name, color='#008ae7D2', fontweight='bold')

    # Plot Counterfactual lines
    for i in range(len(single_points)):
        x_0 = point_to_pred[i]
        x_1 = cf_score
        y_0 = i
        y_1 = cf_pred_x_1
        plt.plot([x_0, x_1], [y_0, y_1], color='k', zorder=0,
                 linewidth=1, alpha=0.15, linestyle='dotted')

    for points, x_value in mulitple_points_chart:
        for origin_point in points:
            x_0 = point_to_pred[origin_point]
            x_1 = x_value
            y_0 = origin_point
            y_1 = np.mean(points)

            plt.plot([x_0, x_1], [y_0, y_1], color='#e0e0e0',
                     zorder=0, linewidth=1, alpha=0.5)

    if save_path is not None:
        plt.savefig(save_path, bbox_inches='tight')
